datafusion/expr/src/expr_schema.rs

```diff
@@ -499,6 +455,76 @@ impl ExprSchemable for Expr {
     }
 }
 
+impl Expr {
+    /// Common method for window functions that applies type coercion
+    /// to all arguments of the window function to check if it matches
+    /// its signature.
+    ///
+    /// If successful, this method returns the data type and
+    /// nullability of the window function's result.
+    ///
+    /// Otherwise, returns an error if there's a type mismatch between
+    /// the window function's signature and the provided arguments.
+    fn data_type_and_nullable_with_window_function(
+        &self,
+        schema: &dyn ExprSchema,
+        window_function: &WindowFunction,
+    ) -> Result<(DataType, bool)> {
```

Comment:
Extracted a common method to handle type coercion for all window function types (built-in, udaf and udwf) which is then reused by methods:,
- `data_type_and_nullable`,
- `get_type` and,
- `nullable`